examples/dpi/dpi.go

```diff
@@ -0,0 +1,382 @@
+}
+
+// Create new counters for new handler
+func (lc localCounters) Copy() interface{} {
+	var newlc localCounters
+	// Clones has the same id
+	id := lc.handlerId
+	newlc.handlerId = id
+	newlc.allowedCounterPtr = &allowedPktsCount[id]
+	newlc.readCounterPtr = &readPktsCount[id]
+	newlc.blockedCounterPtr = &blockedPktsCount[id]
+	return newlc
+}
+
+func (lc localCounters) Delete() {
+}
+
+func filterByRegexp(pkt *packet.Packet, context flow.UserContext) bool {
+	cnt := context.(localCounters)
+	numRead := cnt.readCounterPtr
+	numAllowed := cnt.allowedCounterPtr
+	numBlocked := cnt.blockedCounterPtr
+
+	atomic.AddUint64(numRead, 1)
+	data, err := extractData(pkt)
+	if err != nil {
+		fmt.Println("WARNING:", err, ", drop packet")
+		return false
+	}
+
+	accept := false
+	block := false
+	for _, rule := range rules {
+		result := rule.Re.Match(data)
+		if !result {
+			continue
+		}
+		if rule.Allow {
+			accept = true
+		} else {
+			accept = false
+			block = true
+		}
+	}
+	if accept {
+		atomic.AddUint64(numAllowed, 1)
+	}
+	if !accept && block {
+		atomic.AddUint64(numBlocked, 1)
+	}
+	return accept
+}
+
+func splitBy5Tuple(pkt *packet.Packet, context flow.UserContext) uint {
+	h := fnv.New64a()
+	ip4, ip6, _ := pkt.ParseAllKnownL3()
+	if ip4 != nil {
+		pkt.ParseL4ForIPv4()
+	} else if ip6 != nil {
+		pkt.ParseL4ForIPv6()
+	} else {
+		// Other protocols not supported
+		return totalNumFlows - 1
+	}
+	if ip4 != nil {
+		if ip4.NextProtoID != common.TCPNumber && ip4.NextProtoID != common.UDPNumber {
+			return totalNumFlows - 1
+		}
+		h.Write([]byte{ip4.NextProtoID})
+		buf := new(bytes.Buffer)
+		CheckFatal(binary.Write(buf, binary.LittleEndian, ip4.SrcAddr))
+		h.Write(buf.Bytes())
+		CheckFatal(binary.Write(buf, binary.LittleEndian, ip4.DstAddr))
+		h.Write(buf.Bytes())
+	} else if ip6 != nil {
+		if ip6.Proto != common.TCPNumber && ip6.Proto != common.UDPNumber {
+			return totalNumFlows - 1
+		}
+		binary.Write(h, binary.BigEndian, ip6.Proto)
+		h.Write(ip6.SrcAddr[:])
+		h.Write(ip6.DstAddr[:])
+	}
+
+	binary.Write(h, binary.BigEndian, pkt.GetTCPNoCheck().SrcPort)
+	binary.Write(h, binary.BigEndian, pkt.GetTCPNoCheck().DstPort)
+
+	hash := uint(h.Sum64())
+	return hash % numFlows
+}
+
+func extractData(pkt *packet.Packet) ([]byte, error) {
+	pktLen := pkt.GetPacketSegmentLen()
+	pktStartAddr := pkt.StartAtOffset(0)
+	pktBytes := (*[1 << 30]byte)(pktStartAddr)[:pktLen]
+	ok := pkt.ParseData()
+	if ok == -1 {
+		return nil, fmt.Errorf("cannot extract packet data")
+	}
+
+	hdrsLen := uintptr(pkt.Data) - uintptr(pktStartAddr)
+	return pktBytes[hdrsLen:], nil
+}
+
+func getRulesFromFile(filename string) ([]rule, error) {
+	f, err := ioutil.ReadFile(filename)
+	if err != nil {
+		return nil, err
+	}
+
+	rules = make([]rule, 0)
+	if err := json.Unmarshal(f, &rules); err != nil {
+		return nil, err
+	}
+	for i := 0; i < len(rules); i++ {
+		rules[i].Re = regexp.MustCompile(rules[i].Regexp)
+	}
+	return rules, nil
+}
+
+type rule struct {
```

Comment:
Split example into several files. Create separate directory. Create package rules which will be responsible only for rules handling. Move the common functions to common and left only the logic to the main package.